packages/@kragle/ide/src/tools/node-bindings-editor/use-layout.tsx

```diff
@@ -81,6 +81,277 @@ function calculateLayout(document: SceneDocument): Layout {
+  readonly inputs: Array<string>;
+  /**
+   * Map<columnfrom-columnto,  Set<nodeIDs>>
+   */
+  readonly connections: Map<string, Set<string>>;
+  /**
+   * Array<[strength_of_group, groupID]>
+   */
+  group?: Array<[string, number]>;
+}
+
+function calculateLayout2(document: SceneDocument): Layout {
+  /*
+  let nodeid = document.getRootNodeId()
+  let nodeJson = document.getNode(nodeid!)
+  let childids = Object.values(nodeJson!.slots)
+  let schema = document.nodeDefinitions.get(nodeJson!.type)!.schema
+  let inputNames = Object.keys(schema.inputs)
+  let outputNames = Object.keys(schema.outputs)
+  let binding = nodeJson!.inputs[inputNames[0]]
+  */
+
+  // input initiating
+  let allnodes: Map<string, nodesBox> = new Map();
+  let root: nodesBox = {
+    column: 0,
+    row: [0, -1],
+    ID: document.getRootNodeId()!,
+    inputs: new Array(),
+    connections: new Map(),
+  };
+  allnodes.set(document.getRootNodeId()!, root);
+
+  // input unzip
+  let children: Array<string> = Object.values(document.getNode(root.ID)!.slots);
+  let grandchildren: Array<string> = [];
+  let depth: number = 1;
+  do {
+    for (let c of children) {
+      let nodeJson = document.getNode(c);
+      let inputNames = Object.keys(
+        document.nodeDefinitions.get(nodeJson!.type)!.schema.inputs
+      );
+      let inputIDs: Array<string> = new Array();
+      for (let i of inputNames) {
+        let binding = nodeJson!.inputs[i];
+        if (!binding) {
+          continue;
+        }
+        if (binding.type !== "node-output") {
+          continue;
+        }
+        inputIDs.push(binding.nodeId);
+      }
+      let child: nodesBox = {
+        column: depth,
+        row: [-1, -1],
+        ID: c,
+        inputs: inputIDs,
+        connections: new Map(),
+      };
+      allnodes.set(c, child);
+      let grandchildrencheck = Object.values(nodeJson!.slots);
+      if (grandchildrencheck.length == 0) {
+        continue;
+      }
+      grandchildren = grandchildren.concat(grandchildrencheck);
+    }
+    children = [];
+    if (grandchildren.length > 0) {
+      children = grandchildren;
+      grandchildren = new Array();
+      depth += 1;
+    }
+  } while (children.length > 0);
+
+  // connections find
+  let gap = "-";
+  for (let n of allnodes.values()) {
+    if (n.inputs.length == 0) {
+      continue;
+    }
+    for (let c of n.inputs) {
+      let column_from = allnodes.get(c)!.column;
+      let column_too = n.column;
+      let connection = column_from.toString() + gap + column_too.toString();
+      if (!n.connections.has(connection)) {
+        n.connections.set(connection, new Set());
+      }
+      n.connections.get(connection)!.add(c);
+    }
+  }
+
+  // grouping sort initiating
+  /**
+   * Map<columnfrom-columnto, Set<nodeID>>
+   */
+  let groups: Map<string, Set<string>> = new Map();
+  for (let n of allnodes.values()) {
+    if (n.connections.size == 0) {
+      continue;
+    }
+    for (let c of n.connections) {
```

Comment:
In JavaScript/TypeScript gibt es eine sogenannte [destructuring assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) syntax. Damit kannst du mehrere Variablen deklarieren und sie in einer einzigen Zeile mit den Werten aus einem Array, Tuple, Object, oder Record Elemente fuellen.
```ts
function f(tuple: [string, number]) {
  let [x, y] = tuple; // Deklariert die Variablen `x: string` und `y: number` und weist ihnen die Werte aus dem tuple zu.
}
```

Die `for of` Schleife iteriert ueber ein `Iterable<[K, V]>`, also ueber ein iterable von tuples. Das laesst sich so kombinieren:
```ts
for (let [columnFrom, columnTo] of n.connections) { ... }
```
Dann musst du nicht innerhalb der Schleife auf `c[0]` zugreifen.